src/node.toit

```diff
@@ -0,0 +1,267 @@
+
+See $PartialConcreteNode and $PartialDynamicNode for concrete implementations.
+*/
+abstract class PartialNode extends Node:
+  indentation/string := ""
+
+  can-be-standalone -> bool: return true
+
+/**
+A $PartialNode where the name of the partial is known at compile time.
+
+Example:
+```
+template:
+  {{>text}}
+partials:
+  text: "Hello, world!"
+output:
+  "Hello, world!"
+```
+*/
+class PartialConcreteNode extends PartialNode:
+  name/string
+
+  constructor .name:
+
+  accept visitor/Visitor -> any:
+    return visitor.visit-partial-concrete this
+
+  stringify -> string:
+    return "Partial: $name\n  Indentation: '$indentation'"
+
+/**
+A $PartialNode where the name of the partial is determined at runtime.
+
+Example:
+```
+data:
+  { dynamic: 'content' }
+template:
+  "{{>*dynamic}}"
+partials:
+  { content: 'Hello, world!' }
+output:
+  "Hello, world!"
+```
+
+This node is part of the dynamic-names extension.
+*/
+class PartialDynamicNode extends PartialNode:
+  partial-field/string
+
+  constructor .partial-field:
+
+  accept visitor/Visitor -> any:
+    return visitor.visit-partial-dynamic this
+
+  stringify -> string:
+    return "PartialDynamic: $partial-field\n Indentation: '$indentation'"
+
+/**
+A node that represents a block tag.
+
+Block tags are used to define a block of content that can be overridden
+  by a partial inheritance node.
+
+Example:
+```
+{{$name}}...{{/name}}
+```
+
+Block tags are part of the inheritance extension.
+See $PartialInheritanceNode.
+*/
+class BlockNode extends Node implements ContainerNode:
+  name/string
+  children/List ::= []  // Will be filled by the parser.
+  indentation/string := ""
+
+  constructor .name:
+
+  can-be-standalone -> bool: return true
+
+  add-child node/Node --strict/bool:
+    children.add node
+
+  accept visitor/Visitor -> any:
+    return visitor.visit-block this
+
+  stringify -> string:
+    return "BlockTag: $name (Indentation: '$indentation')"
+
+/**
+A node that represents a partial inheritance.
+
+Partial inheritance is an extension to the mustache syntax that allows
+  a template to inherit from another template.
+
+Example:
+```
+template:
+  {{<super}}{{$title}}sub template title{{/title}}{{/super}}
+partials:
+  super: "...{{$title}}Default title{{/title}}..."
+output:
+  "...sub template title..."
+```
+
+Also see $BlockNode.
+This node is part of the inheritance extension.
+*/
+class PartialInheritanceNode extends PartialNode implements ContainerNode:
+  name/string
+  overridden/Map ::= {:}  // From string to BlockNode.
+
+  constructor .name:
+
+  add-child node/Node --strict/bool:
+    if node is BlockNode:
+      block-node := node as BlockNode
+      if strict and overridden.contains block-node.name:
+        throw "Block tag already overridden: $block-node.name"
+      overridden[block-node.name] = block-node
+    else if strict:
+      throw "Only block tags are allowed in partial inheritance nodes."
```

Comment:
Sometimes you have a dot at the end of strings you throw. Sometimes you don't. I would probably avoid the dots, but some of these are a bit special in that they are almost messages to the user, not just exceptions.